docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,260 @@
+
+Example of the environment variable being turned on:
+
+```
+export AZURE_OS_CACHING=true
+```
+
+The controller will maintain a timestamp in each AzureMachinePool, and when the current time is the chosen interval ahead or more, the controller will perform the caching. Since the current controller manager requeues all objects every ten minutes by default the objects will be requeued shortly after its due time to be recached. This is because typically we expect to cache every 24 hours and it is very unexpected that this won't be frequent enough considering normal patch rates. 
+
+Example of how the timestamp will be maintained in the AzureMachinePool:
+
+```yaml
+status:
+  lastPrototype: "2023-06-12T23:14:55Z"
+```
+
+When the process is started it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a compute image gallery image, delete the snapshot, and then configure the AzureMachinePool specs to use that compute image gallery image. After, it will store the current time as its timestamp. As a note, for the first implementation of this feature we will require the user to also use a compute image gallery image.
+
+Diagram of the Node OS Caching Process:
+
+![Figure 1](./images/node-os-image-cache.png)
+
+As for why the healthy node has to be shut down while creating a snapshot of it, if it isn’t shut down first then pods can be scheduled as the snapshot is taken which will cause some dangerous states in terms of how it exists after being utilized by the AzureMachinePools.
+
+In terms of how a healthy node would be selected, there is already state data present on each AzureMachinePoolMachine which we can use and is listed in the examples below. An ideal node would be one which is running and healthy. Whichever node has been running and healthy for the longest amount of time should be chosen as it’s the most overall stable. This means that for AzureMachinePoolMachines we will take the node with the earliest creation time from metadata.creationTimestamp and has latestModelApplied : true present if it is an AzureMachinePoolMachine since this means that any user changes to the OS image have been rolled out already on this node. As the prototype is always from a successfully healthy and working node the image is always known to be working before being chosen for replication. In terms of knowing if the node os image has had updates that are not user made (so like automatic kernel patches), for the first development of this feature we won't know but all nodes in a cluster are configured to regularly automatically apply security patches to system packages.
```

Comment:
consider breaking down the "Implementation Details" section into subsections to replace the "In terms of" transitions. This could be a "Node Selection" subsection.

Reply:
Done!